src/gtest/test_noteencryption.cpp

```diff
@@ -0,0 +1,170 @@
+#include <gtest/gtest.h>
+#include "sodium.h"
+
+#include <stdexcept>
+
+#include "zcash/NoteEncryption.hpp"
+#include "zcash/prf.h"
+
+class TestNoteDecryption : public ZCNoteDecryption {
+public:
+    TestNoteDecryption(uint256 sk_enc) : ZCNoteDecryption(sk_enc) {}
+
+    void change_pk_enc(uint256 to) {
+        pk_enc = to;
+    }
+};
+
+TEST(noteencryption, api)
+{
+    uint256 sk_enc = ZCNoteEncryption::generate_privkey(uint256S("21035d60bc1983e37950ce4803418a8fb33ea68d5b937ca382ecbae7564d6a77"));
+    uint256 pk_enc = ZCNoteEncryption::generate_pubkey(sk_enc);
+
+    ZCNoteEncryption b = ZCNoteEncryption(uint256());
+
+    boost::array<unsigned char, 216> message;
+    for (unsigned char i = 0; i < 216; i++) {
+        // Fill the message with dummy data
+        message[i] = (unsigned char) i;
+    }
+
+    for (int i = 0; i < 255; i++) {
+        auto ciphertext = b.encrypt(pk_enc, message);
+
+        {
+            ZCNoteDecryption decrypter(sk_enc);
+
+            // Test decryption
+            auto plaintext = decrypter.decrypt(ciphertext, b.get_epk(), uint256(), i);
+            ASSERT_TRUE(plaintext == message);
+
+            // Test wrong nonce
+            ASSERT_THROW(decrypter.decrypt(ciphertext, b.get_epk(), uint256(), (i == 0) ? 1 : (i - 1)), std::runtime_error);
+        
+            // Test wrong ephemeral key
+            ASSERT_THROW(decrypter.decrypt(ciphertext, ZCNoteEncryption::generate_privkey(uint256()), uint256(), i), std::runtime_error);
```

Comment:
I want it to be a valid public key (instead of just a scalar) since `crypto_scalarmult()` could have some sort of extra checks to make sure `epk` is a valid curve point (I'm not sure if it actually does.)